paper2remarkable/utils.py

```diff
@@ -188,6 +192,63 @@ def upload_to_remarkable(filepath, remarkable_dir="/", rmapi_path="rmapi"):
     logger.info("Upload successful.")
 
 
+def list_usb_dir(guid):
+    url = REMARKABLE_USB_URL + "/documents/"
+    if guid:
+        url += guid + "/"
+    response = requests.post(url)
+    return response.json()
+
+
+def crawl_usb_dirs(remarkable_dir):
+    parts = [p for p in remarkable_dir.split("/") if p]
+    guid = None
+    while len(parts) > 0:
+        part = parts.pop(0)
+        data = list_usb_dir(guid)
+        guid = None
+        for item in data:
+            if item["VissibleName"] == part:
+                guid = item["ID"]
+        if not guid:
+            raise RemarkableError(
+                "Directory %s does not exist on reMarkable" % remarkable_dir
+            )
+    # necessary because the last listed dir is the one uploaded to
```

Comment:
I'm aware of the race condition here where a user either has multiple `p2r` processes running at the same time, and the last listed dir creates conflicts (and subsequently files are uploaded to the wrong dir). This seems like it could also happen if they were using the web interface itself.

I make the `list` and `upload` calls right after each other so hopefully the impact should be minimized. However, I am not sure of how to avoid it entirely.

Reply:
Thanks for calling this out, I think it's a good to know about but maybe not very likely in practice. A fairly simple solution could be a creating a file lock, but we can leave that for a potential PR.